s2/gen.go

```diff
@@ -0,0 +1,1590 @@
+	JBE(LabelRef(name + "move_129through256"))
+
+	if avx {
+		JMP(LabelRef(name + "avxUnaligned"))
+	} else {
+		if false {
+			// Don't check length for now.
+			Label(name + "forward")
+			CMPQ(length, U32(2048))
+			JLS(LabelRef(name + "move_256through2048"))
+
+			genMemMove(name+"fallback", dst, src, length, end)
+		} else {
+			JMP(LabelRef(name + "move_256through2048"))
+		}
+	}
+	/*
+			// If REP MOVSB isn't fast, don't use it
+			// FIXME: internal∕cpu·X86+const_offsetX86HasERMS(SB)
+			// CMPB(U8(1), U8(1)) // enhanced REP MOVSB/STOSB
+			JMP(LabelRef(name + "fwdBy8"))
+
+			// Check alignment
+			MOVL(src.As32(), AX.As32())
+			ORL(dst.As32(), AX.As32())
+			TESTL(U32(7), AX.As32())
+			JEQ(LabelRef(name + "fwdBy8"))
+
+			// Do 1 byte at a time
+			// MOVQ(length, CX)
+			// FIXME:
+			// REP;	MOVSB
+			JMP(end)
+
+		Label(name + "fwdBy8")
+		// Do 8 bytes at a time
+		MOVQ(length, CX)
+		SHRQ(U8(3), CX)
+		ANDQ(U8(7), length)
+		// FIXME:
+		//REP;	MOVSQ
+		JMP(LabelRef(name + "tail"))
+
+		Label(name + "back")
+
+		//check overlap
+		MOVQ(src, CX)
+		ADDQ(length, CX)
+		CMPQ(CX, dst)
+		JLS(LabelRef(name + "forward"))
+
+		//whole thing backwards has
+		//adjusted addresses
+
+		ADDQ(length, dst)
+		ADDQ(length, src)
+		STD()
+
+		//
+		//  copy
+		 //
+		MOVQ(length, CX)
+		SHRQ(U8(3), CX)
+		ANDQ(U8(7), length)
+
+		SUBQ(U8(8), dst)
+		SUBQ(U8(8), src)
+		// FIXME:
+		//REP;	MOVSQ
+
+		// FIXME:
+		//CLD()
+
+		ADDQ(U8(8), dst)
+		ADDQ(U8(8), src)
+		SUBQ(length, dst)
+		SUBQ(length, src)
+		JMP(LabelRef(name + "tail"))
+	*/
+
+	Label(name + "move_1or2")
+	MOVB(Mem{Base: src}, AX.As8())
+	MOVB(Mem{Base: src, Disp: -1, Index: length, Scale: 1}, CX.As8())
+	MOVB(AX.As8(), Mem{Base: dst})
+	MOVB(CX.As8(), Mem{Base: dst, Disp: -1, Index: length, Scale: 1})
+	JMP(end)
+
+	Label(name + "move_4")
+	MOVL(Mem{Base: src}, AX.As32())
+	MOVL(AX.As32(), Mem{Base: dst})
+	JMP(end)
+
+	Label(name + "move_3")
+	MOVW(Mem{Base: src}, AX.As16())
+	MOVB(Mem{Base: src, Disp: 2}, CX.As8())
+	MOVW(AX.As16(), Mem{Base: dst})
+	MOVB(CX.As8(), Mem{Base: dst, Disp: 2})
+	JMP(end)
+
+	Label(name + "move_5through7")
+	MOVL(Mem{Base: src}, AX.As32())
+	MOVL(Mem{Base: src, Disp: -4, Index: length, Scale: 1}, CX.As32())
+	MOVL(AX.As32(), Mem{Base: dst})
+	MOVL(CX.As32(), Mem{Base: dst, Disp: -4, Index: length, Scale: 1})
+	JMP(end)
+
+	Label(name + "move_8")
+	// We need a separate case for 8 to make sure we write pointers atomically.
+	MOVQ(Mem{Base: src}, AX)
+	MOVQ(AX, Mem{Base: dst})
+	JMP(end)
+
+	Label(name + "move_9through16")
+	MOVQ(Mem{Base: src}, AX)
+	MOVQ(Mem{Base: src, Disp: -8, Index: length, Scale: 1}, CX)
+	MOVQ(AX, Mem{Base: dst})
+	MOVQ(CX, Mem{Base: dst, Disp: -8, Index: length, Scale: 1})
+	JMP(end)
+
+	Label(name + "move_17through32")
+	X0, X1, X2, X3, X4, X5, X6, X7 := XMM(), XMM(), XMM(), XMM(), XMM(), XMM(), XMM(), XMM()
+	X8, X9, X10, X11, X12, X13, X14, X15 := XMM(), XMM(), XMM(), XMM(), XMM(), XMM(), XMM(), XMM()
```

Comment:
Consider allocating these in an array instead?

```go
X := make([]VecVirtual, 16)
for i := 0; i < 16; i++ {
	X[i] = XMM()
}
```

I had intended to make this even easier: https://github.com/mmcloughlin/avo/issues/44

Reply:
Yeah, but converting old code. This is not about looking nice for now.